rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/History.scala

```diff
@@ -0,0 +1,375 @@
+          Applicative[F].pure((currentKey.dropRight(affix.size.toInt), init, None)).map(_.asLeft)
+
+        case (currentKey, init :+ Skip(affix, _), Some(Skip(eaffix, ep))) => // merge skips
+          rehash(currentKey.dropRight(affix.size.toInt), init, Skip(affix ++ eaffix, ep), Nil)
+            .map(_.asRight)
+
+        case (keyInit :+ keyLast, init :+ Node(p), None) => // handle pointer block
+          pointerBlockStore.get(p).flatMap {
+            case Some(pointerBlock) if pointerBlock.countNonEmpty == 2 =>
+              // pointerBlock contains 1 more element, find it, wrap in skip and re-balance
+              def getOther =
+                pointerBlock
+                  .updated((toInt(keyLast), EmptyTrie) :: Nil)
+                  .toVector
+                  .zipWithIndex
+                  .filter(v => v._1 != EmptyTrie)
+                  .head
+
+              val (other, idx) = getOther
+              Applicative[F].pure((keyInit, init, Some(pointTo(toByte(idx), other)))).map(_.asLeft)
+            case Some(pb) => // pb contains 3+ elements, drop one, rehash
+              val updated = pb.updated((toInt(keyLast), EmptyTrie) :: Nil)
+              storePointerBlock(updated)
+                .flatMap(node => rehash(keyInit, init, node, Nil).map(_.asRight))
+            case None => Sync[F].raiseError(new RuntimeException("malformed trie"))
+          }
+
+        case (key, rest @ _ :+ (_: Node), Some(s: Skip)) =>
+          rehash(key, rest, s, Nil).map(_.asRight)
+
+        case (Nil, Vector(), Some(t)) => Applicative[F].pure(t :: Nil).map(_.asRight)
+
+        case (Nil, Vector(), None) => Applicative[F].pure(EmptyTrie :: Nil).map(_.asRight)
+      }
+    FlatMap[F].tailRecM((p1, r, a))(go)
+  }
+
+  def process(actions: List[HistoryAction[K]]): F[History[F, K]] =
+    // TODO this is an intermediate step to reproduce all the trie behavior
+    // will evolve to a fold based implementation with partial tries
+    actions
+      .foldLeftM(this.root) {
+        case (currentRoot, InsertAction(k, value)) =>
+          val currentPath = asBytes(k)
+          for {
+            traverseResult <- traverseTrie(currentPath, currentRoot)
+            (_, triePath)  = traverseResult
+            elements       <- rebalanceInsert(Leaf(value), currentPath, triePath)
+            _              <- historyStore.put(elements)
+          } yield Trie.hash(elements.last)
+
+        case (currentRoot, DeleteAction(k)) =>
+          val currentPath = asBytes(k)
+          for {
+            traverseResult <- traverseTrie(currentPath, currentRoot)
+            (_, triePath)  = traverseResult
+            elements <- triePath match {
+                         case TriePath(elems :+ (_: Leaf), None, path) if currentPath == path =>
+                           rebalanceDelete(currentPath, elems, None)
+                         case _ =>
+                           Applicative[F].pure[List[Trie]](Nil)
```

Comment:
Shouldn't this case require the path to end at an `Empty` node, and raise an error if it doesn't? Or at least log.warn? After all, in such a case, somebody was able to address the middle of the tree using the 32-byte key passed through public API, or the implementation has a bug...

Reply:
I assumed that it's not a bug to try delete an element that does not exist.

let me ponder if it makes sense to push warn here

Reply:
> not a bug to try delete an element that does not exist.

The non-existence should be known based on the traverse result and then we can make `delete` expect the deleted node to be a Leaf.